Sources/SwiftKueryPostgreSQL/PostgreSQLConnection.swift

```diff
@@ -513,4 +542,46 @@ public class PostgreSQLConnection: Connection {
       }
       return postgresQuery
     }
+    
+    private func lockStateLock() {
+        _ = stateLock.wait(timeout: DispatchTime.distantFuture)
+    }
+    
+    private func unlockStateLock() {
+        stateLock.signal()
+    }
+    
+    func setState(_ newState: ConnectionState) {
+        lockStateLock()
+        if state == .fetchingResultSet {
+            currentResultFetcher = nil
+        }
+        state = newState
+        unlockStateLock()
+    }
+    
+    func setUpForRunningQuery() -> String? {
+        lockStateLock()
+
+        switch state {
+        case .runningQuery:
+            unlockStateLock()
+            return "The connection is in the middle of running a query"
+            
+        case .fetchingResultSet:
+            currentResultFetcher?.hasMoreRows = false
+            unlockStateLock()
+            clearResult(nil, connection: self)
+            lockStateLock()
```

Comment:
I think there's a potential problem here.  Another thread could get in to `setupForRunningQuery()` in between the unlock and lock calls.  The code has to be like this because the lock used is not recursive (and would otherwise result in deadlock).  I wonder if we could be using an `NSRecusiveLock` instead?